lib/diagnostics_channel.js

```diff
@@ -0,0 +1,292 @@
+'use strict';
+
+const {
+  ArrayPrototypeIncludes,
+  ArrayPrototypePush,
+  ArrayPrototypeSplice,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+  Promise,
+  PromiseResolve,
+  RegExp,
+  Symbol
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+const { triggerUncaughtException } = internalBinding('errors');
+
+const { WeakReference } = internalBinding('util');
+
+function alwaysMatch() {
+  return true;
+}
+
+function strictEqualMatch(string) {
+  return function(value) {
+    return value === string;
+  };
+}
+
+function regexpMatch(regexp) {
+  return function(value) {
+    return regexp.test(value);
+  };
+}
+
+function makeSubscriberFor(matcher, onMessage) {
+  if (typeof onMessage !== 'function' && typeof matcher !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE('onMessage', ['function'], onMessage);
+  }
+
+  switch (typeof matcher) {
+    case 'function':
+      if (typeof onMessage === 'function') {
+        return [matcher, onMessage];
+      } else if (typeof onMessage === 'undefined') {
+        return [alwaysMatch, matcher];
+      }
+      break;
+    case 'string':
+    case 'symbol':
+      return [strictEqualMatch(matcher), onMessage];
+    case 'object':
+      if (matcher instanceof RegExp) {
+        return [regexpMatch(matcher), onMessage];
+      }
+      break;
+  }
+
+  throw new ERR_INVALID_ARG_TYPE('matcher',
+                                 ['string', 'symbol', 'function', 'RegExp'],
+                                 matcher);
+}
+
+class AsyncIterableChannel {
+  constructor(channel) {
+    this.channel = channel;
+    this.events = [];
+    this.waiting = [];
+
+    this.subscriber = (message) => {
+      const resolve = this.waiting.shift();
+      if (resolve) {
+        return resolve({
+          value: message,
+          done: false
+        });
+      }
+
+      this.events.push(message);
+    };
+
+    channel.subscribe(this.subscriber);
```

Comment:
do we care if this is not the expected type / `subscribe` method?

Reply:
Should be impossible to access and/or construct without a proper channel, so I think no.